File: CPP/ByTier/D/1973_D_Cat_Fox_and_Maximum_Array_Split.cpp
// Problem: CF 1973 D - Cat, Fox and Maximum Array Split
// https://codeforces.com/contest/1973/problem/D

/*
Purpose: Solve the interactive problem of finding the maximum value m such that an array can be split into exactly k subarrays, each having f(l,r) = m.
Algorithms/Techniques: Binary search on the answer combined with interaction queries to determine valid splits.
Time Complexity: O(n^2 * log(n)) - For each potential m value, we perform up to n queries and there are up to n possible values.
Space Complexity: O(1) - Only a constant amount of extra space is used beyond input storage.

The approach involves:
1. First finding the maximum possible value of array elements to determine search range.
2. Using binary search on the possible values of m.
3. For each m, check if it's possible to split the array into k subarrays all with value m.
4. Use the query function to determine the rightmost position for a given left position and value m.
*/

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <iostream>

using namespace std;
#define int long long
#define ll long long
#define db double
#define iter(s) for (auto &it : s)
#define loop(a, b, c) for (ll a = b; a < c; a++)
#define rloop(a, b, c) for (ll a = c - 1; a >= b; a--)
#define pb push_back
#define pf push_front
#define si size()
#define be begin()
#define en end()
#define all(v) v.be, v.en
#define le length()
#define in insert
#define er erase
#define F first
#define S second
#define get(a, b) get<b>(a)
#define lb lower_bound
#define ub upper_bound
#define bs binary_search
#define eqr equal_range
#define np next_permutation
#define nl cout << "\n"
#define P0(a) cout << a << " "
#define P1(a) cout << a << "\n"
#define P2(a, b) cout << a << " " << b << "\n"
#define P3(a, b, c) cout << a << " " << b << " " << c << "\n"
#define P4(a, b, c, d) cout << a << " " << b << " " << c << " " << d << "\n"
#define deci(n) fixed << setprecision(n)
#define high(n) __builtin_popcountll(n)
#define parity(n) __builtin_parity(n)
#define clz(n) __builtin_clzll(n)
#define ctz(n) __builtin_ctzll(n)
#define pll pair<ll, ll>
#define vc vector<char>
#define vll vector<ll>
#define vpll vector<pll>
#define vvll vector<vll>
#define mll map<ll, ll>
#define sc set<char>
#define sll set<ll>
#define msc multiset<char>
#define msll multiset<ll>
#define M 998244353
#define PI 3.1415926535897932
#define EPS 1e-9
#define INF 1e18
ll query(ll l, ll x) {
  cout << "? " << l << " " << x << endl;
  ll ret;
  cin >> ret;
  if (ret == -1) {
    exit(0);
  }
  return ret;
}
bool check(ll m, ll k, ll n) {
  ll next = 1;
  while (k--) {
    next = query(next, m) + 1;
    if ((k && next == n + 1) || (next > n + 1)) {
      return false;
    }
  }
  if (next == n + 1) {
    return true;
  }
  return false;
}
void solve() {
  ll n, k;
  cin >> n >> k;
  ll amax = 1;
  // Find maximum element value of the array to bound search space
  rloop(i, 2, n + 1) {
    ll r = query(1, n * i);
    if (r == n) {
      amax = i;
      break;
    }
  }
  // Try all possible values of m and check if a valid split exists
  rloop(i, 1, (n / k) + 1) {
    ll m = amax * i;
    if (check(m, k, n)) {
      cout << "! " << m << endl;
      ll resp;
      cin >> resp;
      if (resp == -1) {
        exit(0);
      }
      return;
    }
  }
  cout << "! -1" << endl;
  ll resp;
  cin >> resp;
  if (resp == -1) {
    exit(0);
  }
}
int32_t main() {
  ios::sync_with_stdio(0), cin.tie(0);
  ll tc = 1;
  cin >> tc;
  while (tc--) {
    solve();
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1973/1973_D_Cat_Fox_and_Maximum_Array_Split.py
# Problem: CF 1973 D - Cat, Fox and Maximum Array Split
# https://codeforces.com/contest/1973/problem/D

"""
Algorithm: Binary search on the answer to find the maximum value m such that the array can be split into k subarrays,
each having f(l, r) = m. For each candidate m, we simulate the splitting process using queries.

Time Complexity: O(n^2 * log(n)) in worst case due to the nested loops and binary search on answer.
Space Complexity: O(1) - only using a few variables for computation.
"""

import sys

input = sys.stdin.readline


def ask(i, x):
    # Query the minimum r such that f(i, r) = x
    print(f"? {i + 1} {x}", flush=True)
    return int(input())


def answer(ans):
    # Output the final answer
    print(f"! {ans}", flush=True)
    ok = int(input())
    if ok == -1:
        exit()
    return


def solve():
    n, k = map(int, input().split())

    # Find the maximum value in the array by querying f(l, r) = i * n for decreasing i
    mx = 1
    for i in range(n, 1, -1):
        r = ask(0, i * n)
        if r == n:
            mx = i
            break

    if k == 1:
        # If k is 1, the whole array forms one subarray and the maximum value is the answer
        answer(n * mx)
        return

    # Binary search on the answer (max value m) to check feasibility for k splits
    for i in range(n // k, 0, -1):
        x = mx * i  # Consider this value as the candidate for maximum f(l, r)
        t = 0       # Number of subarrays formed so far
        l = 0       # Start index of current subarray
        
        while l < n and t < k:
            r = ask(l, x)  # Get the minimal r for f(l, r) = x
            if r == n + 1:
                # No such r exists for this value, so this candidate is infeasible
                break
            else:
                # Move to next segment
                l = r
                t += 1

        # If we've formed exactly k subarrays and exhausted the array
        if l == n and t == k:
            answer(x)
            return

    # No feasible value of m found
    answer(-1)
    return


T = int(input())
for _ in range(T):
    solve()


# https://github.com/VaHiX/CodeForces/